### PR TITLE
fix(http): Dynamicaly call the global fetch implementation

### LINKED ISSUE #57527: Allow late instrumentation of `globalThis.fetch`
### Which @angular/* package(s) are relevant/related to the feature request?

common

### Description

I am working on the [Datadog Browser SDK](https://github.com/DataDog/browser-sdk/), and we have customers having [issues](https://github.com/DataDog/browser-sdk/issues/2911) when using [`withFetch()`](https://angular.dev/api/common/http/withFetch).

Our RUM SDK, like many other RUM solutions, is replacing the `fetch` global with an instrumented version (see [our code][1], [Sentry code][2], [OpenTelemetry code][3]). This works fine most of the time, but because Angular 18 is [storing a reference of the `fetch` global internally][4], users willing to use `withFetch()` **have** to load and initialize our SDK before starting their Angular app, else Angular will use the native `fetch` instead of our instrumented version. This is not ideal as it forces the Angular app to wait for our SDK to be loaded and initialized before starting.

[1]: https://github.com/DataDog/browser-sdk/blob/e6f87a87c8f77fc2bb8f10747e1afe10f1e418f0/packages/core/src/browser/fetchObservable.ts#L53
[2]: https://github.com/getsentry/sentry-javascript/blob/e6861cfd08addb1bd1b661067af539fec9e699f9/packages/utils/src/instrument/fetch.ts#L49
[3]: https://github.com/open-telemetry/opentelemetry-js/blob/2ca24594144253afa2bf7ee172bbfe1a1c58969d/experimental/packages/opentelemetry-instrumentation-fetch/src/fetch.ts#L480
[4]: https://github.com/angular/angular/blob/7a99815146eb78074aa3ed6db73c6e87042df692/packages/common/http/src/fetch.ts#L57

### Proposed solution

Keep using the global `fetch` variable. This could be done by replacing

```ts
  private readonly fetchImpl =
    inject(FetchFactory, {optional: true})?.fetch ?? fetch.bind(globalThis);
```

with

```ts
  private readonly fetchImpl =
    inject(FetchFactory, {optional: true})?.fetch ?? (...args) => globalThis.fetch(...args)
```

[here](https://github.com/angular/angular/blob/7a99815146eb78074aa3ed6db73c6e87042df692/packages/common/http/src/fetch.ts#L57C12-L57C24)

### Alternatives considered

Ask Angular 18 users to initialize third-party instrumentation libraries before starting their Angular app.

### PR DESCRIPTION
**Pull request by JeanMeche**
Instead of using the reference that existed when `FetchBackend`  was setup.

fixes #57527

### CODE REVIEW

**Review by AndrewKushnir — APPROVED** [review]

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:667743339:BASE:667743384:1724712217488:6b29a823).

**Comment by AndrewKushnir** [comment]
**Caretaker note**: presubmit is "green", this PR is ready for merge.

File: packages/common/http/src/fetch.ts [hunk 1]
```diff
@@ -54,7 +54,7 @@ function getResponseUrl(response: Response): string | null {
 export class FetchBackend implements HttpBackend {
   // We need to bind the native fetch to its context or it will throw an "illegal invocation"
```
Comment (BenoitZugmeyer) [inline packages/common/http/src/fetch.ts:55]: **💬 suggestion:** ‏maybe adjust this comment as you don't `bind` fetch anymore

Comment (JeanMeche) [reply packages/common/http/src/fetch.ts]: Thank you !

**Comment by JeanMeche** [comment]
Latest change was just a comment update. Will require a re-review though.

**Review by BenoitZugmeyer — APPROVED** [review]

**Review by AndrewKushnir — APPROVED** [review]

**Comment by alxhub** [comment]
This PR was merged into the repository by commit 21445a29322d12fadfb2decaea56f14e95878886.

The changes were merged into the following branches: main, 18.2.x

**Comment by JeanMeche** [comment]
See the fixup that explains why the PR made our HTTP tests flacky.

**Comment by mvdluit** [comment]
> The merge was reverted in #57567.
> 
> See the fixup that explains why the PR made our HTTP tests flaky.

I think you meant to mention this PR: https://github.com/angular/angular/pull/57571

File: packages/common/http/test/fetch_spec.ts [hunk 2]
```diff
@@ -416,6 +419,41 @@ describe('FetchBackend', async () => {
...
+      // We dynamicaly change the implementation of fetch
+      const fakeFetch2 = jasmine
+        .createSpy('', () => Promise.resolve(new Response(JSON.stringify({foo: 'baz'}))))
+        .and.callThrough();
+      globalThis.fetch = fakeFetch2;
+      response = await client.get<unknown>('').toPromise();
+      expect(response).toEqual({foo: 'baz'});
+
+      // We need to restore the original fetch implementation, else the tests might become flaky
+      globalThis.fetch = originalFetch;
```
Comment (JoostK) [inline packages/common/http/test/fetch_spec.ts:454]: A try-finally seems appropriate here

File: packages/common/http/src/fetch.ts [hunk 3]
```diff
@@ -52,9 +52,11 @@ function getResponseUrl(response: Response): string | null {
  */
 @Injectable()
 export class FetchBackend implements HttpBackend {
-  // We need to bind the native fetch to its context or it will throw an "illegal invocation"
+  // Note: binding fetch directly causes an "illegal invocation" error
+  // We use an arrow function to always reference the current global implementation of fetch
+  // In case it has been (monkey)patched after the FetchBackend has been created
```
Comment (AndrewKushnir) [inline packages/common/http/src/fetch.ts:57]: I'd propose adding a bit more info here and dropping the first sentence:
```suggestion
  // We use an arrow function to always reference the current global implementation of `fetch`.
  // This is helpful for cases when the global `fetch` implementation is modified by external code,
  // see https://github.com/angular/angular/issues/57527.
```

**Comment by AndrewKushnir** [comment]
This PR was merged into the repository by commit c2892fee58d28ffec0dfeaad6a5d6822c040cf03.

The changes were merged into the following branches: main, 18.2.x

**Pull request closed without merging**